File: gda-java-components/src/main/java/programmingtheiot/gda/app/DeviceDataManager.java
package programmingtheiot.gda.app;

import java.time.OffsetDateTime;
import java.time.temporal.ChronoUnit;
import java.util.logging.Logger;

import programmingtheiot.common.ConfigConst;
import programmingtheiot.common.ConfigUtil;
import programmingtheiot.common.IActuatorDataListener;
import programmingtheiot.common.IDataMessageListener;
import programmingtheiot.common.ResourceNameEnum;
import programmingtheiot.data.ActuatorData;
import programmingtheiot.data.BaseIotData;
import programmingtheiot.data.DataUtil;
import programmingtheiot.data.SensorData;
import programmingtheiot.data.SystemPerformanceData;
import programmingtheiot.gda.connection.CoapServerGateway;
import programmingtheiot.gda.connection.IPersistenceClient;
import programmingtheiot.gda.connection.IPubSubClient;
import programmingtheiot.gda.connection.MqttClientConnector;
import programmingtheiot.gda.system.SystemPerformanceManager;

public class DeviceDataManager implements IDataMessageListener
{
	private static final Logger _Logger =
		Logger.getLogger(DeviceDataManager.class.getName());

	private boolean enableMqttClient = true;
	private boolean enableCoapServer = false;
	private boolean enableCloudClient = false;
	private boolean enablePersistenceClient = false;
	private boolean enableSystemPerf = false;

	private MqttClientConnector mqttClient = null;
	private IPubSubClient cloudClient = null;
	private IPersistenceClient persistenceClient = null;
	private CoapServerGateway coapServer = null;
	private SystemPerformanceManager sysPerfMgr = null;
	private IActuatorDataListener actuatorDataListener = null;

	private ActuatorData latestHumidifierActuatorData = null;
	private ActuatorData latestHumidifierActuatorResponse = null;
	private SensorData latestHumiditySensorData = null;
	private OffsetDateTime latestHumiditySensorTimeStamp = null;

	private boolean handleHumidityChangeOnDevice = false;
	private int lastKnownHumidifierCommand = ConfigConst.OFF_COMMAND;

	private long humidityMaxTimePastThreshold = 300;
	private float nominalHumiditySetting = 40.0f;
	private float triggerHumidifierFloor = 30.0f;
	private float triggerHumidifierCeiling = 50.0f;

	public DeviceDataManager()
	{
		super();

		ConfigUtil configUtil = ConfigUtil.getInstance();

		this.enableMqttClient =
			configUtil.getBoolean(ConfigConst.GATEWAY_DEVICE, ConfigConst.ENABLE_MQTT_CLIENT_KEY);

		this.enableCoapServer =
			configUtil.getBoolean(ConfigConst.GATEWAY_DEVICE, ConfigConst.ENABLE_COAP_SERVER_KEY);

		this.enableCloudClient =
			configUtil.getBoolean(ConfigConst.GATEWAY_DEVICE, ConfigConst.ENABLE_CLOUD_CLIENT_KEY);

		this.enablePersistenceClient =
			configUtil.getBoolean(ConfigConst.GATEWAY_DEVICE, ConfigConst.ENABLE_PERSISTENCE_CLIENT_KEY);

		this.handleHumidityChangeOnDevice =
			configUtil.getBoolean(ConfigConst.GATEWAY_DEVICE, "handleHumidityChangeOnDevice");

		this.humidityMaxTimePastThreshold =
			configUtil.getInteger(ConfigConst.GATEWAY_DEVICE, "humidityMaxTimePastThreshold");

		this.nominalHumiditySetting =
			configUtil.getFloat(ConfigConst.GATEWAY_DEVICE, "nominalHumiditySetting");

		this.triggerHumidifierFloor =
			configUtil.getFloat(ConfigConst.GATEWAY_DEVICE, "triggerHumidifierFloor");

		this.triggerHumidifierCeiling =
			configUtil.getFloat(ConfigConst.GATEWAY_DEVICE, "triggerHumidifierCeiling");

		if (this.humidityMaxTimePastThreshold < 10 || this.humidityMaxTimePastThreshold > 7200) {
			this.humidityMaxTimePastThreshold = 300;
		}

		initManager();
	}

	public DeviceDataManager(
		boolean enableMqttClient,
		boolean enableCoapClient,
		boolean enableCloudClient,
		boolean enableSmtpClient,
		boolean enablePersistenceClient)
	{
		super();

		this.enableMqttClient = enableMqttClient;
		this.enableCoapServer = enableCoapClient;
		this.enableCloudClient = enableCloudClient;
		this.enablePersistenceClient = enablePersistenceClient;

		initConnections();
	}

	@Override
	public boolean handleActuatorCommandResponse(ResourceNameEnum resourceName, ActuatorData data)
	{
		if (data != null) {
			_Logger.info("Handling actuator response: " + data.getName());

			if (data.hasError()) {
				_Logger.warning("Error flag set for ActuatorData instance.");
			}

			this.latestHumidifierActuatorResponse = data;
			handleIncomingDataAnalysis(resourceName, data);

			return true;
		}

		return false;
	}

	@Override
	public boolean handleActuatorCommandRequest(ResourceNameEnum resourceName, ActuatorData data)
	{
		if (data != null) {
			_Logger.info("Handling actuator request command outbound: " + data.getName());
			handleIncomingDataAnalysis(resourceName, data);
			return true;
		}

		return false;
	}

	@Override
	public boolean handleIncomingMessage(ResourceNameEnum resourceName, String msg)
	{
		if (msg != null) {
			_Logger.info("Handling incoming generic message: " + msg);
			return true;
		}

		return false;
	}

	@Override
	public boolean handleSensorMessage(ResourceNameEnum resourceName, SensorData data)
	{
		if (data != null) {
			_Logger.info("Handling sensor message: " + data.getName());

			if (data.hasError()) {
				_Logger.warning("Error flag set for SensorData instance.");
			}

			String jsonData = DataUtil.getInstance().sensorDataToJson(data);
			int qos = ConfigConst.DEFAULT_QOS;

			if (this.enablePersistenceClient && this.persistenceClient != null) {
				this.persistenceClient.storeData(resourceName.getResourceName(), qos, data);
			}

			handleIncomingDataAnalysis(resourceName, data);
			handleUpstreamTransmission(resourceName, jsonData, qos);

			return true;
		}

		return false;
	}

	@Override
	public boolean handleSystemPerformanceMessage(ResourceNameEnum resourceName, SystemPerformanceData data)
	{
		if (data != null) {
			_Logger.info("Handling system performance message: " + data.getName());

			if (data.hasError()) {
				_Logger.warning("Error flag set for SystemPerformanceData instance.");
			}

			String jsonData = DataUtil.getInstance().systemPerformanceDataToJson(data);
			int qos = ConfigConst.DEFAULT_QOS;

			if (this.enablePersistenceClient && this.persistenceClient != null) {
				this.persistenceClient.storeData(resourceName.getResourceName(), qos, data);
			}

			handleUpstreamTransmission(resourceName, jsonData, qos);

			return true;
		}

		return false;
	}

	@Override
	public void setActuatorDataListener(String name, IActuatorDataListener listener)
	{
		if (listener != null) {
			this.actuatorDataListener = listener;
			_Logger.info("Successfully registered ActuatorDataListener.");
		}
	}

	public void startManager()
	{
		if (this.mqttClient != null) {
			if (this.mqttClient.connectClient()) {
				_Logger.info("Successfully connected MQTT client to broker.");
			} else {
				_Logger.severe("Failed to connect MQTT client to broker.");
			}
		}

		if (this.enableCoapServer && this.coapServer != null) {
			if (this.coapServer.startServer()) {
				_Logger.info("CoAP server started successfully via Manager.");
			} else {
				_Logger.severe("Failed to start CoAP server. Check log file for details.");
			}
		}

		if (this.sysPerfMgr != null) {
			this.sysPerfMgr.startManager();
		}
	}

	public void stopManager()
	{
		if (this.sysPerfMgr != null) {
			this.sysPerfMgr.stopManager();
		}

		if (this.enableCoapServer && this.coapServer != null) {
			this.coapServer.stopServer();
		}

		if (this.mqttClient != null) {
			this.mqttClient.unsubscribeFromTopic(ResourceNameEnum.CDA_ACTUATOR_RESPONSE_RESOURCE);
			this.mqttClient.unsubscribeFromTopic(ResourceNameEnum.CDA_SENSOR_MSG_RESOURCE);
			this.mqttClient.unsubscribeFromTopic(ResourceNameEnum.CDA_SYSTEM_PERF_MSG_RESOURCE);
			this.mqttClient.disconnectClient();
		}
	}

	private void handleIncomingDataAnalysis(ResourceNameEnum resource, ActuatorData data)
	{
		_Logger.fine("Analyzing incoming actuator data: " + data.getName());

		if (data.isResponseFlagEnabled()) {
			this.latestHumidifierActuatorResponse = data;
			_Logger.info("Received actuator response from CDA: " + data.getName());
		}
	}

	private void handleIncomingDataAnalysis(ResourceNameEnum resource, SensorData data)
	{
		_Logger.fine("Analyzing incoming sensor data: " + data.getName());

		if (data.getTypeID() == ConfigConst.HUMIDITY_SENSOR_TYPE) {
			handleHumiditySensorAnalysis(resource, data);
		}
	}

	private void handleHumiditySensorAnalysis(ResourceNameEnum resource, SensorData data)
	{
		if (!this.handleHumidityChangeOnDevice) {
			return;
		}

		_Logger.info("Analyzing humidity data from CDA. Value: " + data.getValue());

		boolean isLow = data.getValue() < this.triggerHumidifierFloor;
		boolean isHigh = data.getValue() > this.triggerHumidifierCeiling;

		if (isLow || isHigh) {
			if (this.latestHumiditySensorData == null) {
				this.latestHumiditySensorData = data;
				this.latestHumiditySensorTimeStamp = getDateTimeFromData(data);

				_Logger.info(
					"Humidity threshold crossed. Waiting for confirmation period: " +
					this.humidityMaxTimePastThreshold +
					" seconds.");

				return;
			}

			OffsetDateTime currentTimeStamp = getDateTimeFromData(data);

			long diffSeconds =
				ChronoUnit.SECONDS.between(
					this.latestHumiditySensorTimeStamp,
					currentTimeStamp);

			_Logger.info("Humidity threshold time delta: " + diffSeconds + " seconds.");

			if (diffSeconds >= this.humidityMaxTimePastThreshold) {
				ActuatorData ad = new ActuatorData();
				ad.setName(ConfigConst.HUMIDIFIER_ACTUATOR_NAME);
				ad.setLocationID(data.getLocationID());
				ad.setTypeID(ConfigConst.HUMIDIFIER_ACTUATOR_TYPE);
				ad.setValue(this.nominalHumiditySetting);

				if (isLow) {
					ad.setCommand(ConfigConst.ON_COMMAND);
				} else {
					ad.setCommand(ConfigConst.OFF_COMMAND);
				}

				_Logger.info("Humidity actuation event triggered. Sending command to CDA: " + ad);

				this.lastKnownHumidifierCommand = ad.getCommand();
				this.latestHumidifierActuatorData = ad;

				sendActuatorCommandToCda(ResourceNameEnum.CDA_ACTUATOR_CMD_RESOURCE, ad);

				this.latestHumiditySensorData = null;
				this.latestHumiditySensorTimeStamp = null;
			}
		} else {
			this.latestHumiditySensorData = null;
			this.latestHumiditySensorTimeStamp = null;

			if (this.lastKnownHumidifierCommand == ConfigConst.ON_COMMAND) {
				ActuatorData ad = new ActuatorData();
				ad.setName(ConfigConst.HUMIDIFIER_ACTUATOR_NAME);
				ad.setLocationID(data.getLocationID());
				ad.setTypeID(ConfigConst.HUMIDIFIER_ACTUATOR_TYPE);
				ad.setCommand(ConfigConst.OFF_COMMAND);
				ad.setValue(this.nominalHumiditySetting);

				_Logger.info("Humidity returned to nominal range. Sending humidifier OFF command.");

				this.lastKnownHumidifierCommand = ConfigConst.OFF_COMMAND;
				this.latestHumidifierActuatorData = ad;

				sendActuatorCommandToCda(ResourceNameEnum.CDA_ACTUATOR_CMD_RESOURCE, ad);
			}
		}
	}

	private void sendActuatorCommandToCda(ResourceNameEnum resource, ActuatorData data)
	{
		if (this.actuatorDataListener != null) {
			this.actuatorDataListener.onActuatorDataUpdate(data);
		}

		if (this.enableMqttClient && this.mqttClient != null) {
			String jsonData = DataUtil.getInstance().actuatorDataToJson(data);

			if (this.mqttClient.publishMessage(resource, jsonData, ConfigConst.DEFAULT_QOS)) {
				_Logger.info("Published ActuatorData command from GDA to CDA: " + data.getCommand());
			} else {
				_Logger.warning("Failed to publish ActuatorData command from GDA to CDA: " + data.getCommand());
			}
		}
	}

	private void handleUpstreamTransmission(ResourceNameEnum resource, String jsonData, int qos)
	{
		_Logger.info("TODO: Send JSON data to cloud service: " + resource);
	}

	private OffsetDateTime getDateTimeFromData(BaseIotData data)
	{
		try {
			return OffsetDateTime.parse(data.getTimeStamp());
		} catch (Exception e) {
			_Logger.warning("Failed to parse timestamp. Using current time.");
			return OffsetDateTime.now();
		}
	}

	private void initConnections()
	{
		if (this.enableMqttClient) {
			this.mqttClient = new MqttClientConnector();
			this.mqttClient.setDataMessageListener(this);
		}

		if (this.enableCoapServer) {
			this.coapServer = new CoapServerGateway(this);
		}
	}

	private void initManager()
	{
		ConfigUtil configUtil = ConfigUtil.getInstance();

		this.enableSystemPerf =
			configUtil.getBoolean(ConfigConst.GATEWAY_DEVICE, ConfigConst.ENABLE_SYSTEM_PERF_KEY);

		if (this.enableSystemPerf) {
			this.sysPerfMgr = new SystemPerformanceManager();
			this.sysPerfMgr.setDataMessageListener(this);
		}

		initConnections();
	}
}

File: cda-python-components/programmingtheiot/cda/app/DeviceDataManager.py
#####
#
# DeviceDataManager.py
#

import logging

import programmingtheiot.common.ConfigConst as ConfigConst

from programmingtheiot.cda.connection.CoapClientConnector import CoapClientConnector
from programmingtheiot.cda.connection.CoapServerAdapter import CoapServerAdapter
from programmingtheiot.cda.connection.MqttClientConnector import MqttClientConnector

from programmingtheiot.cda.system.ActuatorAdapterManager import ActuatorAdapterManager
from programmingtheiot.cda.system.SensorAdapterManager import SensorAdapterManager
from programmingtheiot.cda.system.SystemPerformanceManager import SystemPerformanceManager

from programmingtheiot.common.ConfigUtil import ConfigUtil
from programmingtheiot.common.IDataMessageListener import IDataMessageListener
from programmingtheiot.common.ResourceNameEnum import ResourceNameEnum

from programmingtheiot.data.DataUtil import DataUtil
from programmingtheiot.data.ActuatorData import ActuatorData
from programmingtheiot.data.SensorData import SensorData
from programmingtheiot.data.SystemPerformanceData import SystemPerformanceData


class DeviceDataManager(IDataMessageListener):

    def __init__(self, disableAllComms: bool = False):
        self.configUtil = ConfigUtil()

        self.enableSystemPerf = self.configUtil.getBoolean(
            ConfigConst.CONSTRAINED_DEVICE,
            ConfigConst.ENABLE_SYSTEM_PERF_KEY
        )

        self.enableSensing = self.configUtil.getBoolean(
            ConfigConst.CONSTRAINED_DEVICE,
            ConfigConst.ENABLE_SENSING_KEY
        )

        if disableAllComms:
            self.enableMqttClient = False
            self.enableCoapServer = False
            self.enableCoapClient = False
        else:
            self.enableMqttClient = self.configUtil.getBoolean(
                ConfigConst.CONSTRAINED_DEVICE,
                ConfigConst.ENABLE_MQTT_CLIENT_KEY
            )

            self.enableCoapServer = self.configUtil.getBoolean(
                ConfigConst.CONSTRAINED_DEVICE,
                ConfigConst.ENABLE_COAP_SERVER_KEY
            )

            self.enableCoapClient = self.configUtil.getBoolean(
                ConfigConst.CONSTRAINED_DEVICE,
                ConfigConst.ENABLE_COAP_CLIENT_KEY
            )

        self.enableActuation = True

        self.sysPerfMgr = None
        self.sensorAdapterMgr = None
        self.actuatorAdapterMgr = None

        self.mqttClient = None
        self.coapClient = None
        self.coapServer = None

        self.actuatorDataListener: IDataMessageListener = None
        self.actuatorResponseCache = {}

        if self.enableMqttClient:
            self.mqttClient = MqttClientConnector()
            self.mqttClient.setDataMessageListener(self)
            logging.info("MQTT client enabled")

        if self.enableCoapServer:
            self.coapServer = CoapServerAdapter(dataMsgListener=self)
            logging.info("CoAP server enabled")

        if self.enableCoapClient:
            self.coapClient = CoapClientConnector(dataMsgListener=self)
            logging.info("CoAP client enabled")

        if self.enableSystemPerf:
            self.sysPerfMgr = SystemPerformanceManager()
            self.sysPerfMgr.setDataMessageListener(self)
            logging.info("System performance manager enabled")

        if self.enableSensing:
            self.sensorAdapterMgr = SensorAdapterManager()
            self.sensorAdapterMgr.setDataMessageListener(self)
            logging.info("Sensor manager enabled")

        if self.enableActuation:
            self.actuatorAdapterMgr = ActuatorAdapterManager()
            self.actuatorAdapterMgr.setDataMessageListener(self)
            logging.info("Actuator manager enabled")

        self.handleTempChangeOnDevice = self.configUtil.getBoolean(
            ConfigConst.CONSTRAINED_DEVICE,
            ConfigConst.HANDLE_TEMP_CHANGE_ON_DEVICE_KEY
        )

        self.triggerHvacTempFloor = self.configUtil.getFloat(
            ConfigConst.CONSTRAINED_DEVICE,
            ConfigConst.TRIGGER_HVAC_TEMP_FLOOR_KEY,
            18.0
        )

        self.triggerHvacTempCeiling = self.configUtil.getFloat(
            ConfigConst.CONSTRAINED_DEVICE,
            ConfigConst.TRIGGER_HVAC_TEMP_CEILING_KEY,
            20.0
        )

    def setActuatorDataListener(self, name: str, listener: IDataMessageListener):
        if listener is not None:
            self.actuatorDataListener = listener

    def getLatestActuatorDataResponseFromCache(self, name: str = None) -> ActuatorData:
        return self.actuatorResponseCache.get(name) if name else None

    def getLatestSensorDataFromCache(self, name: str = None) -> SensorData:
        pass

    def getLatestSystemPerformanceDataFromCache(self, name: str = None) -> SystemPerformanceData:
        pass

    def handleActuatorCommandMessage(self, data: ActuatorData = None) -> ActuatorData:
        if data:
            logging.info("Processing actuator command message.")
            return self.actuatorAdapterMgr.sendActuatorCommand(data)
        else:
            logging.warning("Received invalid ActuatorData command message. Ignoring.")
            return None

    def handleActuatorCommandResponse(self, data: ActuatorData = None) -> bool:
        if data:
            self.actuatorResponseCache[data.getName()] = data

            msg = DataUtil().actuatorDataToJson(data)

            if self.coapClient:
                self.coapClient.sendPutRequest(
                    ResourceNameEnum.CDA_ACTUATOR_RESPONSE_RESOURCE,
                    payload=msg
                )

            return True

        return False

    def handleIncomingMessage(self, resourceEnum: ResourceNameEnum = None, msg: str = None) -> bool:
        if resourceEnum and msg:
            self._handleIncomingDataAnalysis(msg)
            return True

        return False

    def handleSensorMessage(self, data: SensorData = None) -> bool:
        if data:
            logging.info("Incoming sensor data received.")

            self._handleSensorDataAnalysis(data=data)

            jsonData = DataUtil().sensorDataToJson(data=data)

            self._handleUpstreamTransmission(
                resourceName=ResourceNameEnum.CDA_SENSOR_MSG_RESOURCE,
                msg=jsonData
            )

            return True

        logging.warning("Incoming sensor data is invalid. Ignoring.")
        return False

    def handleSystemPerformanceMessage(self, data: SystemPerformanceData = None) -> bool:
        if data:
            logging.info("Incoming system performance data received.")

            jsonData = DataUtil().systemPerformanceDataToJson(data=data)

            self._handleUpstreamTransmission(
                resourceName=ResourceNameEnum.CDA_SYSTEM_PERF_MSG_RESOURCE,
                msg=jsonData
            )

            return True

        logging.warning("Incoming system performance data is invalid. Ignoring.")
        return False

    def startManager(self):
        logging.info("Starting DeviceDataManager...")

        if self.mqttClient:
            self.mqttClient.connectClient()

        if self.coapServer:
            self.coapServer.startServer()

        if self.sysPerfMgr:
            self.sysPerfMgr.startManager()

        if self.sensorAdapterMgr:
            self.sensorAdapterMgr.startManager()

        logging.info("DeviceDataManager started")

    def stopManager(self):
        logging.info("Stopping DeviceDataManager...")

        if self.mqttClient:
            self.mqttClient.unsubscribeFromTopic(ResourceNameEnum.CDA_ACTUATOR_CMD_RESOURCE)
            self.mqttClient.disconnectClient()

        if self.coapServer:
            self.coapServer.stopServer()

        if self.sysPerfMgr:
            self.sysPerfMgr.stopManager()

        if self.sensorAdapterMgr:
            self.sensorAdapterMgr.stopManager()

        logging.info("DeviceDataManager stopped")

    def _handleIncomingDataAnalysis(self, msg: str = None):
        try:
            dataUtil = DataUtil()
            actuatorData = dataUtil.jsonToActuatorData(msg)

            if actuatorData and self.actuatorDataListener:
                self.actuatorDataListener.onActuatorDataUpdate(actuatorData)

        except Exception as e:
            logging.error(f"Error analyzing incoming actuator payload: {e}")

    def _handleSensorDataAnalysis(self, resource=None, data: SensorData = None):
        if data and self.handleTempChangeOnDevice and data.getTypeID() == ConfigConst.TEMP_SENSOR_TYPE:
            ad = ActuatorData()
            ad.setName(ConfigConst.HVAC_ACTUATOR_NAME)
            ad.setTypeID(ConfigConst.HVAC_ACTUATOR_TYPE)

            if data.getValue() > self.triggerHvacTempCeiling:
                logging.info("Temperature above ceiling. Triggering HVAC actuator.")
                ad.setCommand(ConfigConst.COMMAND_ON)
                ad.setValue(self.triggerHvacTempCeiling)
                self.handleActuatorCommandMessage(ad)

            elif data.getValue() < self.triggerHvacTempFloor:
                logging.info("Temperature below floor. Triggering HVAC actuator.")
                ad.setCommand(ConfigConst.COMMAND_ON)
                ad.setValue(self.triggerHvacTempFloor)
                self.handleActuatorCommandMessage(ad)

    def _handleUpstreamTransmission(self, resourceName: ResourceNameEnum = None, msg: str = None):
        logging.info("Upstream transmission invoked. Checking comm's integration.")

        if not resourceName or not msg:
            logging.warning("Missing resource name or message. Cannot transmit upstream.")
            return

        if self.mqttClient:
            if self.mqttClient.publishMessage(resource=resourceName, msg=msg):
                logging.debug("Published incoming data to resource using MQTT: %s", str(resourceName))
            else:
                logging.warning("Failed to publish incoming data using MQTT: %s", str(resourceName))

        if self.coapClient:
            if self.coapClient.sendPostRequest(resource=resourceName, payload=msg):
                logging.debug("Posted incoming data to resource using CoAP: %s", str(resourceName))
            else:
                logging.warning("Failed to post incoming data using CoAP: %s", str(resourceName))